kernel/src/hil/sensors.rs

```diff
@@ -289,3 +289,35 @@ pub trait DistanceClient {
     ///                If there was an error, this will be `Err(ErrorCode)`.
     fn callback(&self, distance: Result<u32, ErrorCode>);
 }
+
+/// A basic interface for a rain fall sensor
+pub trait RainFallDriver<'a> {
+    fn set_client(&self, client: &'a dyn RainFallClient);
+
+    /// Read the rain fall value from a sensor. The value is returned
+    /// via the `RainFallClient` callback.
+    ///
+    /// - `hours`: the number of hours of rainfall to report. 0
+    /// is the instant rainfall (if supported). Otherwise 1 to 24
+    /// hours are valid values (if supported by the hardware).
+    ///
+    /// This function might return the following errors:
+    /// - `BUSY`: Indicates that the hardware is busy with an existing
+    ///           operation or initialisation/calibration.
+    /// - `NOSUPPORT`: Indicates that this data type isn't supported.
```

Comment:
```suggestion
    /// - `NOSUPPORT`: Indicates that the value of `hours` is not supported.
```

Data type != data value.

One question: why don't create an associated type, `Hours` which guarantees that the inner value is valid for the given hardware? Something like:

```rust
pub trait HoursLike: TryFrom<u8> {}

pub trait RainFallDriver<'a> {
     type Hours = HoursLike;
     
     fn read_rainfall(&self, hours: Self::Hours) -> Result<(), ErrorCode>;
}
```

The proposal assumes that for all hardware, hours cannot be greater than 24, thus it fits in a u8.

The advantages of this approach:

1. Convey the meaning of the value through its type
2. Enforce checking of the raw `u8` at compile time.
3. Instead of checking if `u8` is valid in every peripheral driver, it will be checked by the capsule once, when trying to pass to the underlying peripheral.

Reply:
I'm not sure I follow, what guarantees that the inner value is valid for the given hardware?

Reply:
I agree that it might be a little more self documenting to have an `Hours` that can only be created with valid values. But I don't see how it happens at compile time? [`try_from()`](https://doc.rust-lang.org/std/convert/trait.TryFrom.html) just returns a result that then needs to be checked. I don't really see the advantage. It seems like it just splits error checking so we need to have more `if Err()` checks

Reply:
Your two examples are basically the same code, implemented in `read_rainfall()` one is just a little Rust-ier.

I don't see how the check happens at compile time, it would just be a run time check

Reply:
Hmm... The check itself is done at runtime, but by adding a wrapper type, the check cannot be missed. Let me reformulate this (although in a hyperbolic manner):

1. One adds the HIL and states in the documentation that `hours` must be an `usize` no greater than 24. The check is done in `read_rainfall()`.
2. One adds the capsule and exposes ` read_rainfall()` to the userspace. Everything works as expected.

### Situation 1

3.1. After 12 months, a new function is added to the HIL, which also takes an `hours` parameter. The check is forgotten by the poor developer.
4.1. The capsule exposes the new function to the userspace. However, when a bad `hours` argument is passed from userspace to the peripheral driver that implements the HIL, no error is returned, but instead some garbage value or worse unexpected behaviour.
5.1. The poor developer spends some time debugging, then they realize they forgot to check whether `hours` is no greater than 24. Then, they ask themselves: why is this a `usize` if it can only take 25 possible values? Shouldn't a wrapper struct with a single field be used for improved type safety? Or even better, an enum with 25 variants?

### Situation 2

3.2. After 12 months, a new rainfall sensor is added, but the lazy developer doesn't read documentation and forgets to check whether `hours` is no greater than 24.
4.2. After some testing, garbage value seems to be returned in some cases. After some debugging, the lazy developer realizes they forgot the check. Then, they ask themselves: why is this a `usize` if it can only take 25 possible values? [...] (or even worse, they don't test values greater than 24, but a GitHub is opened a few months latter)

One might say: the underlined situations can be easily mitigated by careful reviews. One may answer: yeah, such as [this](https://github.com/tock/tock/pull/4161) issue.

Now you decide whether it is worth the pain adding an additional layer of abstraction. :)